### PR TITLE
fix: auto-detect EOT token from chat template when eos_token is absent (Gemma)

### LINKED ISSUE #3754: `chat_template` Strategy Never Trains the Gemma Turn Terminator
Credit: @ferno9 for discovering the issue and diagnosing 

### Please check that this issue hasn't been reported before.

- [x] I searched previous [Bug Reports](https://github.com/axolotl-ai-cloud/axolotl/labels/bug) didn't find any similar reports.

### Expected Behavior

When fine-tuning any **Gemma** model (1 / 2 / 3 / 3n / 4) with `type: chat_template`, the per-turn stop token that the Gemma chat template actually emits — `<end_of_turn>` (id **107** on Gemma 1/2, id **106** on Gemma 3/3n) or `<turn|>` (id **106** on Gemma 4) — should be **trained** at the end of each assistant turn (it should receive a real label, honoring `train_on_eos`/`train_on_eot`), so the model learns to stop. This already happens by default for other families (Qwen `<|im_end|>`, Llama-3 `<|eot_id|>`, Mistral `{{ eos_token }}`, …).

### Current behaviour

For Gemma, the turn terminator is **silently masked (`-100`) by default**, so the model is never taught to stop. After fine-tuning, the model produces the correct response. Loss/eval curves look perfectly healthy because the only thing never learned — stopping — is never in the loss.

**Root cause:** `ChatTemplateStrategy` defaults `eot_tokens` to `[tokenizer.eos_token]`, and the EOS/EOT finders key off `tokenizer.eos_token_id`. For Gemma that is `<eos>` (id **1**), but `<eos>` **never appears in the rendered chat** (turns end with `<end_of_turn>`/`<turn|>`), so neither finder matches the turn boundary → the terminator's label stays `-100`.

A warning *is* emitted but is easy to miss and not an error:

```
[WARNING] [axolotl.prompt_strategies.chat_template] EOS token '<eos>' not found in chat_template. Please check if your template/EOS token is correct.
```
The default gemma configs (e.g. Gemma 4) don't include the config flags needed to train the real stop token (the most likely fine-tuning scenario for non-base variants). The override works however:

```
  special_tokens:
    eos_token: "<turn|>"     # "<end_of_turn>" for gemma 2/3/3n
  train_on_eos: turn

```

**Why other models are fine:** their `tokenizer.eos_token` *is* the turn terminator (Qwen → `<|im_end|>`, Llama-3-instruct → `<|eot_id|>`) or the template renders the `{{ eos_token }}` variable. Gemma is the only family where HF keeps `tokenizer.eos_token = <eos>` and exposes the real terminator only via `generation_config.eos_token_id` (e.g. gemma-4-31B-it = `[1, 106, 50]`, gemma-3-4b-it = `[1, 106]`). So the eos-keyed default misses it for **every Gemma generation**.

Relevant code: `src/axolotl/prompt_strategies/chat_template.py`
- `eot_tokens` defaults to `[self.tokenizer.eos_token]` (~L310–317)
- `_validate_eot_and_eos_tokens` warns but does not fix (~L350–374)
- label assignment via `find_first_eos_token` / `find_first_eot_token`, both keyed on token id 1 (~L599–637)

## Steps to reproduce

1. Create a tiny chat dataset `train.jsonl` (2 lines is enough), e.g.:
   ```json
   {"messages":[{"role":"user","content":"Translate to French: The weather is nice today."},{"role":"assistant","content":"Il fait beau aujourd'hui."}]}
   {"messages":[{"role":"user","content":"Translate to French: We met at the library."},{"role":"assistant","content":"Nous nous sommes rencontrés à la bibliothèque."}]}
   ```
2. Use the minimal config below (the official `examples/gemma4/31b-qlora.yaml` reproduces it identically — it sets neither `eot_tokens` nor `special_tokens.eos_token`).
3. Run preprocess with label debug:
   ```bash
   python -m axolotl.cli.preprocess config.yaml --debug 2>&1 | grep -E "not found in chat_template|turn\|>\("
   ```
4. **Observe** the warning and the masked terminator on every assistant turn:
   ```
   ... aujourd'hui .(236761, 236761) <turn|>(-100, 106)     <-- masked, never trained
   ```
   (Same result with `ds_type: json` or `ds_type: parquet`, and with `sample_packing: true` or `false` — each packed sample's `<turn|>` is masked too, so packing is not involved.)

5. **Confirm the workaround** trains it — add to the config:
   ```yaml
   special_tokens:
     eos_token: "<turn|>"   # use "<end_of_turn>" for Gemma 1/2/3/3n
   train_on_eos: turn
   ```
   Re-run step 3 → warning is gone and the terminator is now trained:
   ```
   ... la bibliothèque .(236761, 236761) <turn|>(106, 106)   <-- trained
   ```

### Config yaml

```yaml
base_model: google/gemma-4-31B-it
chat_template: gemma4
datasets:
  - path: ./train.jsonl
    ds_type: json
    type: chat_template
    field_messages: messages
adapter: lora
lora_r: 16
lora_alpha: 16
lora_target_modules: 'model.language_model.layers.[\d]+.(_checkpoint_wrapped_module.)?(mlp|self_attn).(up|down|gate|q|k|v|o)_proj'
sequence_len: 2048
micro_batch_size: 1
num_epochs: 2
# No special_tokens / eot_tokens set -> reproduces the bug (matches examples/gemma4/31b-qlora.yaml)
```

### Possible solution

If the intended behavior is to reduce code complexity the example configs could be updated to note the overrides are required for training on EOS otherwise.

More robust fix (mirrors how Unsloth and LLaMA-Factory handle it):

1. **Derive the trainable terminator from the rendered chat template** instead of `tokenizer.eos_token`. Render the active template on a dummy assistant turn and read the token(s) emitted **after** the assistant content; default `eot_tokens` to that token when the user hasn't set it. The tokenizer resolves the correct id automatically, so this needs **no hardcoded ids/strings** and handles `<end_of_turn>`=107 (Gemma 1/2) vs 106 (Gemma 3/3n) vs `<turn|>` (Gemma 4) transparently. It also generalizes to any other model with an `eos_token` ≠ terminator split.
   - Optional secondary cross-check: `generation_config.eos_token_id` (lists the terminator on Gemma 3/3n/4). Do **not** rely on it alone — Gemma 1 and gemma-2-9b/27b-it omit it, and Gemma 3 *base* also lists `[1,106]`, so it's not a reliable base-vs-instruct signal.

2. **Gate on chat-template presence** This is essentially free (the `chat_template` strategy only runs when a template is in play) and avoids mis-handling non-chat Gemma derivatives. 

3. **Keep the existing override** (`eot_tokens` / `special_tokens.eos_token`) as the always-wins escape hatch — just make the default correct so it isn't required.

4. **(Optional, recommended) also override `tokenizer.eos_token` → terminator** (cf. LLaMA-Factory `replace_eos`, Unsloth `map_eos_token`) so the **saved** adapter/model stops at inference too. The train-side and export-side fixes are independent; doing both prevents "answer then never stops" at train *and* serve time.

A minimal regression test: preprocess a Gemma chat sample and assert the terminator token id receives a non-`-100` label (and the `<eos> not found` warning is gone).

### Which Operating Systems are you using?

- [x] Linux
- [ ] macOS
- [ ] Windows

### Python Version

3.12

### axolotl branch-commit

main/4d6490b

### Acknowledgements

- [x] My issue title is concise, descriptive, and in title casing.
- [x] I have searched the existing issues to make sure this bug has not been reported yet.
- [x] I am using the latest version of axolotl.
- [x] I have provided enough information for the maintainers to reproduce and diagnose the issue.

### PR DESCRIPTION
**Pull request by JSap0914**
## Description

Auto-detect the actual end-of-turn (EOT) token from the chat template when
`tokenizer.eos_token` is absent from it. This fixes Gemma models where the
turn terminator (`<end_of_turn>` / `<turn|>`) was silently masked and never
trained.

## Motivation and Context

Closes #3754

For every Gemma generation (1/2/3/3n/4), `tokenizer.eos_token = "<eos>"` (id 1),
but the chat template emits `<end_of_turn>` (Gemma 1/2/3/3n, id 107/106) or
`<turn|>` (Gemma 4, id 106) as the per-turn stop marker.

The old code defaulted `eot_tokens = [tokenizer.eos_token]` and cached token
id 1. `find_first_eot_token` never found a match because id 1 never appears in
rendered Gemma sequences, leaving every turn terminator with label -100. The
model was never trained to stop, but loss/eval curves look fine because the
untrained token never contributed to the loss.

A warning was already emitted, but not acted upon.

PR #3756 (open) adds an `eot_tokens` workaround to example YAML configs and
improves the warning. This PR fixes the root cause in `chat_template.py` so
Gemma works out of the box without any config change.

### `src/axolotl/prompt_strategies/chat_template.py`

- When `_validate_eot_and_eos_tokens` detects that `eos_token` is absent from
  the template, call `_detect_eot_from_template()` instead of just warning.
- `_detect_eot_from_template()`: renders the template with a sentinel assistant
  turn, locates the sentinel token-ids in the output, then collects the
  non-whitespace non-EOS tokens that immediately follow. These are the actual
  turn terminator(s). Updates `self.eot_tokens` with them.
- If detection fails, clears `self.eot_tokens` (correct: stop searching for the
  wrong id) and logs a warning with explicit instructions.
- Explicit `eot_tokens` set by the user always win; the new logic only triggers
  when the default `[eos_token]` is in use and that token is absent from the
  template.
- Templates that reference `eos_token` as a Jinja variable (Llama-3, ChatML,
  Mistral, …) are completely unaffected.

## Verification

```
pytest tests/prompt_strategies/test_gemma_eot_autodetect.py -v
# 5 passed
```

All tests use a mock tokenizer and are fully offline (no GPU / model download).
No existing test was edited.

```
pytest tests/prompt_strategies/test_chat_template_utils.py \
       tests/prompt_strategies/test_jinja_template_analyzer.py \
       tests/prompt_strategies/test_synthetic.py \
       tests/integrations/test_kd_chat_template.py -v
# 33 passed
```

## Summary by CodeRabbit

* **Bug Fixes**
  * Improved detection of chat turn terminators when the template uses a different end marker than the tokenizer’s EOS token.
  * Added fallback behavior that clearly warns when a terminator can’t be identified, preventing incorrect training on turn endings.
  * Preserved user-provided terminator settings so they aren’t overwritten by automatic detection.

* **Tests**
  * Added coverage for Gemma-like chat templates, EOS mismatch cases, explicit terminator overrides, and terminator position detection.

### CODE REVIEW

**Comment by thad0ctor** [comment]
I'd recommend double checking your transformers version meets or exceeds the pinned version in your environment.

```python
full_ids = list(self.tokenizer.apply_chat_template(..., tokenize=True))
```

This assumes a list of ids, but on transformers v5 `return_dict` defaults to `True`, so `apply_chat_template` returns a `BatchEncoding` and `list(...)` gives the keys instead:

```
['input_ids', 'attention_mask']
```

This failed silently when I tested

**Comment by NanoCode012** [comment]
I think I mentioned this elsewhere. There's no specific code that you can gate for Gemma chat template (besides to update the yaml which was already done), so this change impacts all chat template parsing.

As some templates do not have EOT, this can be breaking change, so holding off on this. Did you verify with other templates?
